mathics/eval/tensors.py

```diff
@@ -76,6 +74,88 @@ def get_dimensions(expr, head=None):
         return [len(expr.elements)] + sub
 
 
+def to_std_sparse_array(sparse_array, evaluation: Evaluation):
+    if sparse_array.elements[2] == Integer0:
+        return sparse_array
+    else:
+        return Expression(
+            SymbolSparseArray, Expression(SymbolNormal, sparse_array)
+        ).evaluate(evaluation)
+
+
+def unpack_outer(item, rest_lists, current, level: int, const_etc: tuple):
+    """
+    Recursively unpacks lists to evaluate outer product.
+    ------------------------------------
+
+    Unlike direct products, outer (tensor) products require traversing the
+    lowest level of each list, hence we recursively unpacking lists until
+    the lowest level is reached.
+
+    Parameters:
+
+    ``item``: the current item to be unpacked (if not at lowest level), or joined
+    to current (if at lowest level)
+
+    ``rest_lists``: the rest of lists to be unpacked
+
+    ``current``: the current lowest level elements
+
+    ``level``: the current level (unused yet, will be used in
+    ``Outer[f_, lists__, n_]`` in the future)
+
+    ``const_etc``: a tuple of functions used in unpacking, remains constant
+    throughout the recursion.
+
+    Format of ``const_etc``:
+
+    ```
+    (
+        cond_next_list,  # return True/False to unpack the next list/this list at next level
+        get_elements,  # get elements of list, tuple, ListExpression, etc.
+        apply_head,  # e.g. lambda elements: Expression(head, *elements)
+        apply_f,  # e.g. lambda current: Expression(f, *current)
+        join_elem,  # join current lowest level elements (i.e. current) with a new one
+        if_nested,  # True for result as nested list, False for result as flattened list
+        evaluation,  # evaluation: Evaluation
+    )
+    ```
+    """
+    (

```

Comment:
I would think that passing a dictionary would be more "Pythonic" (explicit) than passing a magic tuple as an argument. One way to do this is

```
def unpack_outer(item, rest_lists, current, level: int, const_etc: dict):
      locals().update(const_etc)
```